R/calltrace.R

```diff
@@ -0,0 +1,212 @@
+#' Capture a call trace.
+#'
+#' A call trace captures sequence of calls that lead to the current function,
+#' sometimes called the call stack. Because of lazy evaluation, the call
+#' stack in R is actually a tree, which the print method of this object will
+#' reveal.
+#'
+#' @param top_env If non-null, this will be used to automatically trim the
+#'   call trace to only show calls after the last occurence of this enviroment.
+#'   This is needed when the code that calls `calltree()` is run indirectly,
+#'   for example in tests, or inside an RMarkdown document.
+#' @examples
+#' f <- function() g()
+#' g <- function() h()
+#' h <- function() calltrace()
+#'
+#' # When no lazy evaluation is involved the calltrack is linear
+#' # (i.e. every call has only one child)
+#' f()
+#'
+#' # Lazy evaluation introduces a tree like structure
+#' identity(identity(f()))
+#' identity(try(f()))
+#' try(identity(f()))
+#'
+#' # When printing, you can request to simplify this tree to only show
+#' # the direct sequence of calls that lead to `calltrace()`
+#' x <- try(identity(f()))
+#' x
+#' print(x, simplify = TRUE)
+#'
+#' # With a little cunning you can also use it to capture the
+#' # tree from within a base NSE function
+#' x <- NULL
+#' with(mtcars, {x <<- f(); 10})
+#' x
+#'
+#' # When code is executed indirectly, i.e. via source or within an
+#' # RMarkdown document, you'll tend to get a lot of guff at the beginning
+#' # related to the execution environment:
+#' source(textConnection("f()"), echo = TRUE)
+#'
+#' # To automatically strip this off, pass `top_env = globalenv()`
+#' # That will automatically trim of calls prior to the last appearance
+#' # of the global environment on the stack
+#' h <- function() calltrace(globalenv())
+#' source(textConnection("f()"), echo = TRUE)
+#' @export
+calltrace <- function(top_env = NULL) {
+  calls <- sys.calls()
+  parents <- sys.parents()
+  envs <- map(sys.frames(), env_label)
+
+  funs <- map(1:sys.nframe(), sys.function)
+  refs <- map(funs, attr, "srcref")
+
+  trace <- new_calltrace(calls, parents, envs, refs)
+  trace <- trace_trim_env(trace, top_env)
+  trace <- trace[-length(trace)] # remove call to self
+
+  trace
+}
+
+new_calltrace <- function(calls, parents, envs, refs) {
+  stopifnot(is.list(calls), is.integer(parents), length(calls) == length(parents))
+
+  structure(
+    list(
+      calls = calls,
+      parents = parents,
+      envs = envs,
+      refs = refs
+    ),
+    class = "calltrace"
+  )
+}
+
+# Methods -----------------------------------------------------------------
+
+#' @export
+format.calltrace <- function(x, simplify = FALSE, dir = getwd(), ...) {
+  if (length(x) == 0) {
+    return("\u2588")
```

Comment:
Do you need a non-unicode fallback for this?

Reply:
Oops, yes.